runtime/drive_handler.go

```diff
@@ -15,125 +15,227 @@ package main
-	models "github.com/firecracker-microvm/firecracker-go-sdk/client/models"
+	"github.com/firecracker-microvm/firecracker-go-sdk/client/models"
 
 	"github.com/firecracker-microvm/firecracker-containerd/internal"
+	"github.com/firecracker-microvm/firecracker-containerd/proto"
+	drivemount "github.com/firecracker-microvm/firecracker-containerd/proto/service/drivemount/ttrpc"
 )
 
 const (
-	// fcSectorSize is the sector size of Firecracker
+	// fcSectorSize is the sector size of Firecracker drives
 	fcSectorSize = 512
 )
 
 var (
 	// ErrDrivesExhausted occurs when there are no more drives left to use. This
 	// can happen by calling PatchStubDrive greater than the number of drives.
 	ErrDrivesExhausted = fmt.Errorf("There are no remaining drives to be used")
-
-	// ErrDriveIDNil should never happen, but we safe guard against nil dereferencing
-	ErrDriveIDNil = fmt.Errorf("DriveID of current drive is nil")
 )
 
-// stubDriveHandler is used to manage stub drives.
-type stubDriveHandler struct {
-	RootPath       string
-	stubDriveIndex int64
-	drives         []models.Drive
-	logger         *logrus.Entry
-	mutex          sync.Mutex
+// CreateContainerStubs will create a StubDriveHandler for managing the stub drives
+// of container rootfs drives. The Firecracker drives are hardcoded to be read-write
+// and have no rate limiter configuration.
+func CreateContainerStubs(
+	machineCfg *firecracker.Config,
+	jail jailer,
+	containerCount int,
+	logger *logrus.Entry,
+) (*StubDriveHandler, error) {
+	var containerStubs []*stubDrive
+	for i := 0; i < containerCount; i++ {
+		isWritable := true
+		var rateLimiter *proto.FirecrackerRateLimiter
+		stubFileName := fmt.Sprintf("ctrstub%d", i)
+
+		stubDrive, err := newStubDrive(
+			filepath.Join(jail.JailPath().RootPath(), stubFileName),
+			jail, isWritable, rateLimiter, logger)
+
+		if err != nil {
+			return nil, errors.Wrap(err, "failed to create container stub drive")
+		}
+
+		machineCfg.Drives = append(machineCfg.Drives, models.Drive{
+			DriveID:      firecracker.String(stubDrive.driveID),
+			PathOnHost:   firecracker.String(stubDrive.stubPath),
+			IsReadOnly:   firecracker.Bool(!isWritable),
+			RateLimiter:  rateLimiterFromProto(rateLimiter),
+			IsRootDevice: firecracker.Bool(false),
+		})
+		containerStubs = append(containerStubs, stubDrive)
+	}
+
+	return &StubDriveHandler{
+		freeDrives: containerStubs,
+		usedDrives: make(map[string]*stubDrive),
+	}, nil
 }
 
-// stubDrivesOpt is used to make and modify changes to the stub drives.
-type stubDrivesOpt func(stubDrives []models.Drive) error
+// StubDriveHandler manages a set of stub drives. It currently only supports reserving
+// one of the drives from its set.
+// In the future, it may be expanded to also support recycling a drive to be used again
+// for a different mount.
+type StubDriveHandler struct {
+	freeDrives []*stubDrive
+	// map of id -> stub drive being used by that task
+	usedDrives map[string]*stubDrive
```

Comment:
What is the purpose of this field? I don't really see it used anywhere other than being set

Reply:
I'm just laying some groundwork for the implementation of #243 to build off of, which I am personally okay with since it's not complicating anything at the moment. If there's disagreement with including it even though it technically doesn't need to exist yet, then I can remove for now, let me know.

Reply:
Ah, okay. I was just making sure it was there intentionally. I am okay with leaving it.